cdc/owner/changefeed.go

```diff
@@ -524,37 +537,48 @@ LOOP:
 	return nil
 }
 
+// releaseResources is idempotent.
 func (c *changefeed) releaseResources(ctx cdcContext.Context) {
+	if c.isReleased {
+		return
+	}
 	// Must clean redo manager before calling cancel, otherwise
 	// the manager can be closed internally.
 	c.cleanupRedoManager(ctx)
-
-	if !c.initialized {
-		c.cleanupChangefeedServiceGCSafePoints(ctx)
-		return
-	}
+	c.cleanupChangefeedServiceGCSafePoints(ctx)
 
 	c.cancel()
 	c.cancel = func() {}
-	c.ddlPuller.Close()
-	c.schema = nil
-	c.cleanupChangefeedServiceGCSafePoints(ctx)
-	canceledCtx, cancel := context.WithCancel(context.Background())
-	cancel()
-	// We don't need to wait sink Close, pass a canceled context is ok
-	if err := c.sink.close(canceledCtx); err != nil {
-		log.Warn("owner close sink failed",
-			zap.String("namespace", c.id.Namespace),
-			zap.String("changefeed", c.id.ID),
-			zap.Error(err))
+
+	if c.ddlPuller != nil {
+		c.ddlPuller.Close()
+		c.ddlPuller = nil
+	}
+	c.ddlWg.Wait()
+
+	if c.sink != nil {
+		canceledCtx, cancel := context.WithCancel(context.Background())
+		cancel()
+		// TODO(dongmen): remove ctx from func sink.close(), it is useless.
+		// We don't need to wait sink Close, pass a canceled context is ok
+		if err := c.sink.close(canceledCtx); err != nil {
+			log.Warn("owner close sink failed",
+				zap.String("namespace", c.id.Namespace),
+				zap.String("changefeed", c.id.ID),
+				zap.Error(err))
+		}
+	}
+
+	if c.scheduler != nil {
+		c.scheduler.Close(ctx)
+		c.scheduler = nil
 	}
-	c.wg.Wait()
-	c.scheduler.Close(ctx)
-	c.scheduler = nil
-	c.barriers = nil
 
 	c.cleanupMetrics()
+	c.schema = nil
+	c.barriers = nil
 	c.initialized = false
+	c.isReleased = true
```

Comment:
it looks that `initialized = !isReleased`, so can we only keep one?

Reply:
the have sligthly different, `isReleased` is assigned to `false` [here](https://github.com/pingcap/tiflow/pull/6882/files#diff-953746562776cf8aa41cd4a971df8c546649d2721ae787607aed2069f3b44eb0R387),  which indicated that we need to release this changefeed if an error occurs in `initialize()`.